R/filter.R

```diff
@@ -33,3 +34,22 @@ porcelain_filter_metadata <- function(req, res) {
   req$received_time <- now_utc()
   plumber::forward()
 }
+
+
+porcelain_filter_request_id <- function(logger) {
+  force(logger)
+  function(req, res) {
+    if ("x-request-id" %in% names(req$HEADERS)) {
+      request_id <- req$HEADERS[["x-request-id"]]
+    } else {
+      request_id <- ids::uuid()
+    }
+    req$REQUEST_ID <- request_id
+    res$setHeader("x-request-id", request_id)
+    if (!is.null(logger)) {
+      logger$add_filter(lgr::FilterInject$new(request_id = request_id),
```

Comment:
We chatted about this in person but just putting this down here for future reference. 

We can't just use `porcelain_log_postserialize` as we want to add the request ID into the logger so if a user writes a log call in their API impl it will pick up the request ID.

With plumber we have 4 hooks we can use preroute, postroute, preserialize and posterialize. In preroute we don't have a handle on the "HTTP_X_REQUEST_ID" so we need to do this after the preroute, but by actual code endpoint is run before we get to postroute. So we have to handle this in a filter.

I did some benchmarking on this, by running up the API in a background process with the `add.chatty` package I added in this PR. There isn't a huge amount of variation in timings

With no filter

```
rashton@wpia-dide270:~/projects/porcelain$ ./timeit curl -w "@curl_format.txt" -o /dev/null -s  "http://localhost:8552/?a=3&b=3"
Average: 0.00569371 N:  1000
```

With the filter which creates and removes the object every time
```
rashton@wpia-dide270:~/projects/porcelain$ ./timeit curl -w "@curl_format.txt" -o /dev/null -s  "http://localhost:9120/?a=3&b=3"
Average: 0.0055903 N:  1000
```

With updated filter which just changes the request ID on an the filter object which exists in cache
```
rashton@wpia-dide270:~/projects/porcelain$ ./timeit curl -w "@curl_format.txt" -o /dev/null -s  "http://localhost:9343/?a=3&b=3"
Average: 0.00546013 N:  1000
```

So very similar average times, less different than I was expecting.